utility/service/service.go

```diff
@@ -70,17 +96,102 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
 		return nil, fmt.Errorf("Error admitting relay: %w", err)
 	}
 
-	// TODO: implement Persist Relay
-	// TODO: implement execution
-	// TODO: implement state maintenance
-	// TODO: validate the response from the node?
-	// TODO: (QUESTION) Should we persist SignedRPC?
-	return nil, nil
+	response, err := s.executeRelay(relay)
+	if err != nil {
+		return nil, fmt.Errorf("Error executing relay: %w", err)
+	}
+
+	// DISCUSS: should we validate the response from the node?
+	relayDigest, shouldStore, err := s.hasCollision(relay, response)
+	if err != nil {
+		return nil, fmt.Errorf("Error calculating relay service digest: %w", err)
+	}
+	if !shouldStore {
+		return response, nil
+	}
+
+	height := s.GetBus().GetConsensusModule().CurrentHeight()
+	writeCtx, err := s.GetBus().GetPersistenceModule().NewRWContext(int64(height))
+	if err != nil {
+		return nil, fmt.Errorf("Error getting a write context to update token usage for application %s: %w", relay.Meta.ApplicationAddress, err)
+	}
+	defer writeCtx.Release()
+
+	// DISCUSS: should we extend/use UnitOfWork for updating/retrieving token usage?
+	if err := writeCtx.RecordRelayService(relay.Meta.ApplicationAddress, relayDigest, relay, response); err != nil {
+		return nil, fmt.Errorf("Error recording service proof for application %s: %w", relay.Meta.ApplicationAddress, err)
+	}
+
+	return response, nil
+}
+
+// hasCollision returns:
+//  1. The signed digest of a relay/response pair,
+//  2. Whether there was a collision for the specific chain (i.e. should the service proof be stored for claiming later)
+func (s *servicer) hasCollision(relay *coreTypes.Relay, response *coreTypes.RelayResponse) (digest []byte, collides bool, err error) {
+	relayBytes, err := marshal(relay, response)
+	if err != nil {
+		return nil, false, fmt.Errorf("Error marshalling relay and/or response: %w", err)
+	}
+
+	relayDigest := crypto.SHA3Hash(relayBytes)
+
+	signedDigest := s.sign(relayDigest)
+	response.ServicerSignature = hex.EncodeToString(signedDigest)
+	collision, err := s.hasCollisionOnChain(relayDigest, relay.Meta.RelayChain.Id)
+	if err != nil {
+		return nil, false, fmt.Errorf("Error checking collision for chain %s: %w", relay.Meta.RelayChain.Id, err)
+	}
+
+	return signedDigest, collision, nil
+}
+
+// INCOMPLETE: implement this
+func (s *servicer) sign(bz []byte) []byte {
+	return bz
+}
+
+// INCOMPLETE: implement this
+func (s *servicer) hasCollisionOnChain(digest []byte, relayChainId string) (bool, error) {
+	return false, nil
+}
+
+func marshal(request *coreTypes.Relay, response *coreTypes.RelayResponse) ([]byte, error) {
```

Comment:
s/marshal/marshalWhat?

1. Rename
s/marshal/marshalWhat?

2. Comment

Even though it's an internal helper, add a comment because this functionality is VERY important for the core of the protocol.

3. Refactor

The integrity of the blockchain (i.e. the protobufs and their serialized values in the tree) are the most important piece of the state. Ideally, people should be able to use the protos and implement it on their own. 

Can you create a proto in `relay.proto` that accepts both of these so that the marshaling of the digest is done by `codec.GetCodec().Marshal(RelayReqRes)`. That way we are language and framework agnostic and others, for example, could reimplement v1 in rust using the same protos.

```proto
message RelayReqRes {
	RelayReq
	RelayRes
}
```

Reply:
Thank you for the review. Fixed.